integrations/kube-agent-updater/pkg/controller/statefulset.go

```diff
@@ -0,0 +1,284 @@
+	var (
+		noNewVersionErr *NoNewVersionError
+		maintenanceErr  *MaintenanceNotTriggeredError
+	)
+	switch {
+	case errors.As(err, &noNewVersionErr):
+		// Error contains the detected versions
+		log.Info("StatefulSet is already up-to-date, not updating.", "err", err)
+		return requeueLater, nil
+	case errors.As(err, &maintenanceErr):
+		// Not logging the error because it provides no other information than its type.
+		log.Info("No maintenance triggered, not updating.", "currentVersion", currentVersion)
+		return requeueLater, nil
+	case trace.IsTrustError(err):
+		// Logging as error as image verification should not fail under normal use
+		log.Error(err, "Image verification failed, not updating.")
+		return requeueLater, trace.Wrap(err)
+	case err != nil:
+		log.Error(err, "Unexpected error, not updating.")
+		return requeueLater, trace.Wrap(err)
+	}
+
+	log.Info("Updating podSpec with image", "image", image.String())
+	err = setContainerImageFromPodSpec(&obj.Spec.Template.Spec, teleportContainerName, image.String())
+	if err != nil {
+		return requeueLater, trace.Wrap(err)
+	}
+
+	if err = r.Update(ctx, &obj); err != nil {
+		return requeueNow, trace.Wrap(err)
+	}
+	return requeueLater, nil
+}
+
+// SetupWithManager makes the DeploymentVersionUpdater managed by a ctrl.Manager.
+// Once started, the manager will send Deployment reconciliation requests to the
+// DeploymentVersionUpdater controller.
+func (r *StatefulSetVersionUpdater) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&appsv1.StatefulSet{}).
+		Complete(r)
+}
+
+func (r *StatefulSetVersionUpdater) unblockStatefulSetRollout(ctx context.Context, sts *appsv1.StatefulSet) error {
+	log := ctrllog.FromContext(ctx)
+	verboseLog := ctrllog.FromContext(ctx).V(1)
+
+	if sts.Status.UpdateRevision == "" {
+		return trace.BadParameter("statefulset '%s' UpdateRevision empty", sts.Name)
+	}
+
+	// First we get all pods managed by the StatefulSet
+	stsLabelSelector, err := metav1.LabelSelectorAsSelector(sts.Spec.Selector)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	var managedPodsList v1.PodList
+	listOptions := []kclient.ListOption{kclient.InNamespace(sts.Namespace), kclient.MatchingLabelsSelector{Selector: stsLabelSelector}}
+	err = r.List(ctx, &managedPodsList, listOptions...)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// *v1.Pod implements metav1.Object, it's easier to convert now from
+	// []v1.Pod to []*v1.Pod. e.g. this is required for listNames()
+	managedPods := podListToListOfPods(&managedPodsList)
+
+	verboseLog.Info("statefulset managed pods", "managedPodsList", listNames(managedPods))
+
+	// We filter out malformed pods, not created by the statefulset
+	// We don't want to destroy other pods in case of label conflicts
+	var saneManagedPods []*v1.Pod
+	for _, pod := range managedPods {
+		if (pod.Labels == nil) || (pod.Labels[appsv1.ControllerRevisionHashLabelKey] == "") {
```

Comment:
Nit: You could probably just filter out healthy pods and pods managed by the current controller revision in this loop as well, would make the code a little bit more compact.

Reply:
And on the related note, I would consider factoring out the "get old unhealthy pods" logic into a separate method. Then the flow of this "unblock" method would be much clearer: 1. we get old unhealthy pods and 2. if there are any, we delete them.

Reply:
As this logic was common with the maintenance unhealthy trigger I introduced a generic pod filtering logic through `podutils.Filters`.